File: src/main/java/com/ggemo/va/goingmerry/handler/ClassicFindHandlerInRegistryHandler.java
package com.ggemo.va.goingmerry.handler;

import com.ggemo.va.goingmerry.handler.handleranalyse.impl.ClassicConditionAnalyseResult;
import com.ggemo.va.goingmerry.handler.handlerregistry.ClassicHandlerRegistry;
import com.ggemo.va.handler.OpHandler;

public class ClassicFindHandlerInRegistryHandler
        implements OpHandler<SelectHandlerBusiness.Context<ClassicConditionAnalyseResult>, OpHandler<?, ?>> {
    private final ClassicHandlerRegistry registry;

    public ClassicFindHandlerInRegistryHandler() {
        registry  = ClassicHandlerRegistry.getInstance();
    }

    private static ClassicFindHandlerInRegistryHandler INSTANCE;

    public static ClassicFindHandlerInRegistryHandler getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new ClassicFindHandlerInRegistryHandler();
        }
        return INSTANCE;
    }

    @Override
    public OpHandler<?, ?> handle(SelectHandlerBusiness.Context<ClassicConditionAnalyseResult> req) {
        if (!registry.registered(req.getHandlerClazz())) {
            registry.initRegister(req.getHandlerClazz());
        }
        return registry.findHandler(req.getAnalyseResult(), req.getHandlerClazz());
    }
}

File: src/main/java/com/ggemo/va/goingmerry/handler/ClassicFindHandlerInRegistryHandler.kt
package com.ggemo.va.goingmerry.handler

import com.ggemo.va.goingmerry.handler.handleranalyse.impl.ClassicConditionAnalyseResult
import com.ggemo.va.goingmerry.handler.handlerregistry.ClassicHandlerRegistry
import com.ggemo.va.handler.OpHandler

class ClassicFindHandlerInRegistryHandler
    : OpHandler<SelectHandlerBusiness.Context<ClassicConditionAnalyseResult>, OpHandler<*, *>> {
    private val registry: ClassicHandlerRegistry = ClassicHandlerRegistry.getInstance();

    companion object {
        private val INSTANCE = ClassicFindHandlerInRegistryHandler();

        fun getInstance(): ClassicFindHandlerInRegistryHandler {
            return INSTANCE
        }
    }

    override fun handle(req: SelectHandlerBusiness.Context<ClassicConditionAnalyseResult>): OpHandler<*, *> {
        if (!registry.registered(req.getHandlerClazz())) {
            registry.initRegister(req.getHandlerClazz())
        }
        return registry.findHandler(req.getAnalyseResult(), req.getHandlerClazz())
    }
}